File: app/src/main/java/com/simon/harmonichackernews/SearchFragment.java
package com.simon.harmonichackernews;

import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.google.android.material.button.MaterialButtonToggleGroup;
import com.simon.harmonichackernews.databinding.FragmentSearchBinding;

import java.util.ArrayList;

public class SearchFragment extends Fragment {
    FragmentSearchBinding binding;
    String searchQuery;
    ArrayList<String> tags;
    private final String SHOW_HN = "show_hn";
    private final String ASK_HN = "ask_hn";
    private final String FRONT_PAGE = "front_page";
    private final String STORY = "story";
    private final String COMMENT = "comment";
    private final String POLL = "poll";

    public SearchFragment() {
        super();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        binding = FragmentSearchBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }
    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.getRoot().setClickable(true);
        searchQuery = "";
        tags = new ArrayList<>();
        EditText searchBar  = binding.searchEditText;
        searchBar.requestFocus();
        searchBar.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView textView, int actionId, KeyEvent keyEvent) {
                searchQuery = searchBar.getText().toString();
                return sendResult();
            }
        });
        binding.searchNormalToolbar.setNavigationOnClickListener(this::dismissSearch);
        binding.postSpecialTagToggleGroup.addOnButtonCheckedListener(new MaterialButtonToggleGroup.OnButtonCheckedListener() {
            @Override
            public void onButtonChecked(MaterialButtonToggleGroup group, int checkedId, boolean isChecked) {
                int index = group.indexOfChild(group.findViewById(checkedId));
                switch (index){
                    case 0 -> {
                        updateTag(isChecked, SHOW_HN);
                    }
                    case 1 -> {
                        updateTag(isChecked, ASK_HN);
                    }
                    case 2 -> {
                        updateTag(isChecked, FRONT_PAGE);
                    }
                }
            }
        });

        binding.postTypeToggleGroup.addOnButtonCheckedListener(new MaterialButtonToggleGroup.OnButtonCheckedListener() {
            @Override
            public void onButtonChecked(MaterialButtonToggleGroup group, int checkedId, boolean isChecked) {
                int index = group.indexOfChild(group.findViewById(checkedId));
                switch (index){
                    case 0 -> {
                        updateTag(isChecked, STORY);
                    }
                    case 1 -> {
                        updateTag(isChecked, COMMENT);
                    }
                    case 2 -> {
                        updateTag(isChecked, POLL);
                    }
                }
            }
        });
    }
    private void updateTag(boolean isChecked, String tag){
        if(isChecked){
            tags.add(tag);
        } else {
            tags.remove(tag);
        }
    }
    private boolean sendResult() {
        Bundle result = new Bundle();
        result.putString("query_term", searchQuery);
        result.putStringArrayList("tags", tags);
        getParentFragmentManager().setFragmentResult("search_query", result);
        return true;
    }
    private void dismissSearch(View v){
        searchQuery = "";
        sendResult();
    }
}

File: app/src/main/java/com/simon/harmonichackernews/view/search/SearchFragment.kt
package com.simon.harmonichackernews.view.search

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.core.widget.doAfterTextChanged
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.DividerItemDecoration.VERTICAL
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.transition.MaterialSharedAxis
import com.simon.harmonichackernews.CommentsActivity
import com.simon.harmonichackernews.R
import com.simon.harmonichackernews.StoriesFragment.StoryClickListener
import com.simon.harmonichackernews.adapters.StoryListAdapter
import com.simon.harmonichackernews.data.PostType
import com.simon.harmonichackernews.data.SortType
import com.simon.harmonichackernews.data.Story
import com.simon.harmonichackernews.databinding.FragmentSearchBinding
import com.simon.harmonichackernews.utils.CommentsUtils
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class SearchFragment : Fragment() {
    private lateinit var binding: FragmentSearchBinding
    private val viewModel: SearchViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentSearchBinding.inflate(inflater, container, false)
        enterTransition = MaterialSharedAxis(MaterialSharedAxis.Z, true).setDuration(600)
        returnTransition = MaterialSharedAxis(MaterialSharedAxis.Z, false).setDuration(600)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.searchView.editText.setOnEditorActionListener { _, _, _ -> search() }

        val clickListener: (Story) -> Unit = { story ->
            (requireActivity() as? StoryClickListener)?.openStory(story, 0, false)
        }
        val storyClickListener: (Story) -> Unit = { story ->
            val intent =
                Intent(requireActivity().applicationContext, CommentsActivity::class.java).apply {
                    putExtra(CommentsUtils.EXTRA_ID, story.commentMasterId)
                    putExtra(CommentsUtils.EXTRA_TITLE, story.commentMasterTitle)
                    putExtra(CommentsUtils.EXTRA_URL, story.commentMasterUrl)
                }
            startActivity(intent)
        }
        val adapter = StoryListAdapter(
            requireContext(),
            onLinkClick = clickListener,
            onCommentsCLick = clickListener,
            onCommentStoryClick = storyClickListener,
            onCommentRepliesClick = clickListener,
        )

        with(binding.recyclerView) {
            layoutManager = LinearLayoutManager(context)
            addItemDecoration(DividerItemDecoration(requireContext(), VERTICAL))
            this.adapter = adapter
        }

        viewModel.loading.observe(viewLifecycleOwner) { loading ->
            binding.loadingProgressbar.isVisible = loading
        }
        binding.searchBar.setNavigationOnClickListener { v: View -> dismissSearch(v) }
        viewModel.searchResults.observe(viewLifecycleOwner) { stories ->
            binding.noResultsFoundTextview.isVisible = stories.isEmpty()
            adapter.submitList(stories)
        }

        binding.searchSettings.postTypeChipGroup.setOnCheckedStateChangeListener { _, checkedIds ->
            val types = checkedIds.mapNotNull { id ->
                when (id) {
                    R.id.story_chip -> PostType.STORY
                    R.id.comment_chip -> PostType.COMMENT
                    R.id.poll_chip -> PostType.POLL
                    R.id.show_hn_chip -> PostType.SHOW_HN
                    R.id.ask_hn_chip -> PostType.ASK_HN
                    else -> null
                }
            }
            viewModel.setPostTypes(types.takeUnless { it.isEmpty() } ?: listOf(PostType.STORY))
        }
        binding.searchSettings.sortTypeToggleGrp.addOnButtonCheckedListener { _, checkedId, isChecked ->
            if (isChecked) {
                when (checkedId) {
                    R.id.by_relevance_btn -> viewModel.setSortType(SortType.BY_RELEVANCE)
                    R.id.by_date_btn -> viewModel.setSortType(SortType.BY_DATE)
                }
            }
        }


        binding.searchSettings.frontPageSwitch.setOnCheckedChangeListener { _, isChecked ->
            viewModel.setFrontPage(isChecked)
        }
        binding.searchSettings.usernameEditText.doAfterTextChanged { author ->
            viewModel.setAuthor(author.toString())
        }
        binding.searchSettings.searchBtn.setOnClickListener {
            search()
        }
    }

    private fun search(): Boolean {
        val query = binding.searchView.text.toString()
        binding.searchBar.setText(query)
        viewModel.search(query)
        binding.searchView.hide()
        return false
    }

    private fun dismissSearch(v: View) {
        parentFragmentManager.popBackStack()
    }
}
